processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -98,287 +167,247 @@ public void start()
-          continue;
-        }
-        if (!lookupExtractorFactory.start()) {
-          failedExtractorFactoryMap.put(lookupName, lookupExtractorFactory);
-          continue;
-        }
-        if (null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory)) {
-          // handle race
-          LOGGER.warn("lookup [%s] is not add, another lookup with the same name already exist", lookupName);
-          if (!lookupExtractorFactory.close()) {
-            LOGGER.error("Failed to properly close stale lookup [%s]", lookupExtractorFactory);
-          }
-          continue;
-        }
-        if (lookupSnapshotTaker != null) {
-          lookupSnapshotTaker.takeSnapshot(getAllAsList());
-        }
+      if (!queue.offer(new DropNotice(lookupName), 1, TimeUnit.MILLISECONDS)) {
+        throw new ISE("notice queue add timedout to add [%s] lookup drop notice", lookupName);
       }
-      if (!failedExtractorFactoryMap.isEmpty()) {
-        throw new ISE(
-            "was not able to start the following lookup(s) [%s]",
-            failedExtractorFactoryMap.keySet().toString()
-        );
-      }
-    }
-    finally {
-      startStopLock.readLock().unlock();
+    } catch (InterruptedException ex) {
+      throw new ISE(ex, "failed to add [%s] lookup drop notice", lookupName);
     }
   }
 
-  /**
-   * Add or update a lookup factory
-   *
-   * @param lookupName             The name of the lookup
-   * @param lookupExtractorFactory The factory of the lookup
-   *
-   * @return True if the lookup was updated, false otherwise
-   *
-   * @throws IllegalStateException if start of the factory fails
-   */
-  public boolean updateIfNew(String lookupName, final LookupExtractorFactory lookupExtractorFactory)
+  @Nullable
+  public LookupExtractorFactoryContainer get(String lookupName)
   {
-    boolean update = false;
+    assertStarted();
+
+    startStopLock.readLock().lock();
     try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
+      return lookupMap.get(lookupName);
+    } finally {
+      startStopLock.readLock().unlock();
     }
-    try {
-      assertStarted();
-      LookupExtractorFactory prior = lookupMap.get(lookupName);
-      update = lookupExtractorFactory.replaces(prior);
-      if (update) {
-        if (!lookupExtractorFactory.start()) {
-          throw new ISE("Could not start [%s]", lookupName);
-        }
-        boolean racy;
-        do {
-          if (prior == null) {
-            racy = null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory);
-          } else {
-            racy = !lookupMap.replace(lookupName, prior, lookupExtractorFactory);
-          }
+  }
 
-          if (racy) {
-            prior = lookupMap.get(lookupName);
-            update = lookupExtractorFactory.replaces(prior);
-          }
-        } while (racy && update);
+  public LookupsState getAllLookupsState()
```

Comment:
1) I think using BlockingQueue is better than using park/unpark as "when things are available" is automatically handled and add/remove methods don't need to do anything other than adding things to the queue.

2) I do like `LifecycleLock` , it is great as it takes away the need of repeated code in druid of managing the "started" variable . I think it should be used. However , it is in a separate PR .. do you mind me pulling `LifecycleLock` class into this PR?

3) I took a hard look at this class again to see if explicit synchronization can be completely removed. And, I found in the handling of LoadNotice, it starts a new factory and adds it to the lookupMap . With no synchronization , in a race condition , it will be possible that close() method and LoadNotice run in parallel and end result being that the factory just started by LoadNotice is not closed by the close() code.  This is one race condition I could identify, not sure and there might be more. Having explicit locks provides guarantees and given that this code does not have to be super high performant.. I'm still thinking of maintaining the locks.

@leventov what do you think?